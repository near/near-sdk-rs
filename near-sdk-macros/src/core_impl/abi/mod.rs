#[cfg(feature = "__abi-embed")]
mod abi_embed;
#[cfg(feature = "__abi-embed")]
pub use abi_embed::embed;

#[cfg(feature = "__abi-generate")]
mod abi_generator;
#[cfg(feature = "__abi-generate")]
pub use abi_generator::generate;
