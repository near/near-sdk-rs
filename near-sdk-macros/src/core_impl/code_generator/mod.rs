mod attr_sig_info;
pub use attr_sig_info::*;

mod impl_item_method_info;
pub use impl_item_method_info::*;

mod item_trait_info;
pub use item_trait_info::*;

mod trait_item_method_info;
pub use trait_item_method_info::*;

mod item_impl_info;
pub use item_impl_info::*;

mod item_struct_info;
pub use item_struct_info::*;
