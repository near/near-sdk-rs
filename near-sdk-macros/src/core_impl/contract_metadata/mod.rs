use darling::{ast::NestedMeta, Error, FromMeta};
use proc_macro2::TokenStream;
use quote::quote;

#[derive(FromMeta)]
struct MacroConfig {
    contract_metadata: Option<ContractMetadata>,
}

#[derive(serde::Serialize, Default, FromMeta)]
struct ContractMetadata {
    version: Option<String>,
    link: Option<String>,
    #[darling(multiple, rename = "standard")]
    standards: Vec<Standard>,
}

#[derive(FromMeta, serde::Serialize)]
struct Standard {
    standard: String,
    version: String,
}

impl ContractMetadata {
    fn populate(mut self) -> Self {
        if self.version.is_none() {
            let version = std::env::var("CARGO_PKG_VERSION").unwrap_or(String::from(""));
            if !version.is_empty() {
                self.version = Some(version);
            }
        }

        if self.link.is_none() {
            let repo = std::env::var("CARGO_PKG_REPOSITORY").unwrap_or(String::from(""));
            if !repo.is_empty() {
                self.link = Some(repo);
            }
        }

        // adding nep330 if it is not present
        if self.standards.is_empty()
            || self.standards.iter().all(|s| s.standard.to_ascii_lowercase() != "nep330")
        {
            self.standards
                .push(Standard { standard: "nep330".to_string(), version: "1.1.0".to_string() });
        }

        self
    }
}

/// Allows for the injection of the contract source metadata infomation into the contract code as
/// a constant.
pub(crate) fn contract_source_metadata_const(attr: proc_macro::TokenStream) -> TokenStream {
    if attr.to_string().is_empty() {
        let metadata = serde_json::to_string(&ContractMetadata::default().populate())
            .expect("ContractMetadata implements Serialize");

        return quote! {
           pub const CONTRACT_SOURCE_METADATA: &'static str = #metadata;
        };
    }

    let attr_args = match NestedMeta::parse_meta_list(attr.into()) {
        Ok(v) => v,
        Err(e) => {
            return Error::from(e).write_errors();
        }
    };

    let args = match MacroConfig::from_list(&attr_args) {
        Ok(v) => v,
        Err(e) => {
            return e.write_errors();
        }
    };

    let metadata = serde_json::to_string(
        &args
            .contract_metadata
            .expect("Attribute input must be present given standard was followed")
            .populate(),
    )
    .expect("ContractMetadata implements Serialize");

    quote! {
        const CONTRACT_SOURCE_METADATA: &'static str = #metadata;
    }
}
