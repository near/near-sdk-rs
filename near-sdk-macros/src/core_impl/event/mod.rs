use proc_macro::TokenStream;
use proc_macro2::Span;
use quote::quote;
use syn::{parse_quote, ItemEnum, LitStr};

/// this function is used to inject serialization macros and the `near_sdk::EventMetadata` macro.
/// In addition, this function extracts the event's `standard` value and injects it as a constant to be used by
/// the `near_sdk::EventMetadata` derive macro
pub(crate) fn near_events(attr: TokenStream, item: TokenStream) -> TokenStream {
    // get standard from attr args
    let standard = get_standard_arg(&syn::parse_macro_input!(attr as syn::AttributeArgs));
    if standard.is_none() {
        return TokenStream::from(
            syn::Error::new(
                Span::call_site(),
                "Near events must have a `standard` value as an argument for `event_json` in the `near_bindgen` arguments. The value must be a string literal, e.g. \"nep999\", \"mintbase-marketplace\".",
            )
            .to_compile_error(),
        );
    }

    if let Ok(mut input) = syn::parse::<ItemEnum>(item) {
        let name = &input.ident;
        let standard_name = format!("{}_event_standard", name);
        let standard_ident = syn::Ident::new(&standard_name, Span::call_site());
        // NearEvent Macro handles implementation
        input
            .attrs
            .push(parse_quote! (#[derive(near_sdk::serde::Serialize, near_sdk::EventMetadata)]));
        input.attrs.push(parse_quote! (#[serde(crate="near_sdk::serde")]));
        input.attrs.push(parse_quote! (#[serde(tag = "event", content = "data")]));
        input.attrs.push(parse_quote! (#[serde(rename_all = "snake_case")]));

        TokenStream::from(quote! {
            const #standard_ident: &'static str = #standard;
            #input
        })
    } else {
        TokenStream::from(
            syn::Error::new(
                Span::call_site(),
                "`#[near_bindgen(event_json(standard = \"nepXXX\"))]` can only be used as an attribute on enums.",
            )
            .to_compile_error(),
        )
    }
}

/// This function returns the `version` value from `#[event_version("x.x.x")]`.
/// used by `near_sdk::EventMetadata`
pub(crate) fn get_event_version(var: &syn::Variant) -> Option<LitStr> {
    for attr in var.attrs.iter() {
        if attr.path.is_ident("event_version") {
            return attr.parse_args::<LitStr>().ok();
        }
    }
    None
}

/// this function returns the `standard` value from `#[near_bindgen(event_json(standard = "nepXXX"))]`
fn get_standard_arg(args: &[syn::NestedMeta]) -> Option<LitStr> {
    let mut standard: Option<LitStr> = None;
    for arg in args.iter() {
        if let syn::NestedMeta::Meta(syn::Meta::List(syn::MetaList { path, nested, .. })) = arg {
            if path.is_ident("event_json") {
                for event_arg in nested.iter() {
                    if let syn::NestedMeta::Meta(syn::Meta::NameValue(syn::MetaNameValue {
                        path,
                        lit: syn::Lit::Str(value),
                        ..
                    })) = event_arg
                    {
                        if path.is_ident("standard") {
                            standard = Some(value.to_owned());
                            break;
                        }
                    }
                }
            }
        }
    }
    standard
}
