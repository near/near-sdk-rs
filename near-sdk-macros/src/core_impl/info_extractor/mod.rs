mod serializer_attr;
pub use serializer_attr::SerializerAttr;

mod arg_info;
pub use arg_info::{ArgInfo, BindgenArgType};

mod attr_sig_info;
pub use attr_sig_info::AttrSigInfo;

mod impl_item_method_info;
pub use impl_item_method_info::ImplItemMethodInfo;

mod trait_item_method_info;
pub use trait_item_method_info::*;

mod item_trait_info;
pub use item_trait_info::ItemTraitInfo;

mod item_impl_info;

mod init_attr;
pub use init_attr::InitAttr;

pub use item_impl_info::ItemImplInfo;

/// Type of serialization we use.
#[derive(PartialEq, Eq)]
#[allow(clippy::upper_case_acronyms)]
pub enum SerializerType {
    JSON,
    Borsh,
}

/// Type of the method.
#[derive(PartialEq, Eq)]
pub enum MethodType {
    Regular,
    View,
    Init,
    InitIgnoreState,
}

/// Whether the input struct is used for serialization or deserialization.
#[derive(PartialEq, Eq)]
pub enum InputStructType {
    Serialization,
    Deserialization,
}
