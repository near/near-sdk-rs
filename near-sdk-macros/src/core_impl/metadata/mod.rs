pub(crate) mod metadata_generator;
pub(crate) mod metadata_visitor;
