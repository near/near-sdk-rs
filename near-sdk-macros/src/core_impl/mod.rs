mod code_generator;
mod info_extractor;
mod metadata;
pub use code_generator::*;
pub use info_extractor::*;
pub use metadata::metadata_visitor::MetadataVisitor;
