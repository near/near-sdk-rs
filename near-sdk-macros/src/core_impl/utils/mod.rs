use proc_macro2::{Group, Span, TokenStream as TokenStream2, TokenTree};
use quote::quote;
use syn::spanned::Spanned;
use syn::token::{And, Mut};
use syn::{GenericArgument, Path, PathArguments, Signature, Type};

#[cfg(test)]
pub mod test_helpers;

/// Checks whether the given path is literally "Result".
/// Note that it won't match a fully qualified name `core::result::Result` or a type alias like
/// `type StringResult = Result<String, String>`.
pub(crate) fn path_is_result(path: &Path) -> bool {
    path.leading_colon.is_none()
        && path.segments.len() == 1
        && path.segments.iter().next().unwrap().ident == "Result"
}

/// Equivalent to `path_is_result` except that it works on `Type` values.
pub(crate) fn type_is_result(ty: &Type) -> bool {
    match ty {
        Type::Path(type_path) if type_path.qself.is_none() => path_is_result(&type_path.path),
        _ => false,
    }
}

/// Extracts the Ok type from a `Result` type.
///
/// For example, given `Result<String, u8>` type it will return `String` type.
pub(crate) fn extract_ok_type(ty: &Type) -> Option<&Type> {
    match ty {
        Type::Path(type_path) if type_path.qself.is_none() && path_is_result(&type_path.path) => {
            // Get the first segment of the path (there should be only one, in fact: "Result"):
            let type_params = &type_path.path.segments.first()?.arguments;
            // We are interested in the first angle-bracketed param responsible for Ok type ("<String, _>"):
            let generic_arg = match type_params {
                PathArguments::AngleBracketed(params) => Some(params.args.first()?),
                _ => None,
            }?;
            // This argument must be a type:
            match generic_arg {
                GenericArgument::Type(ty) => Some(ty),
                _ => None,
            }
        }
        _ => None,
    }
}

/// Checks whether the given path is literally "Vec".
/// Note that it won't match a fully qualified name `std::vec::Vec` or a type alias like
/// `type MyVec = Vec<String>`.
#[cfg(feature = "__abi-generate")]
fn path_is_vec(path: &Path) -> bool {
    path.leading_colon.is_none()
        && path.segments.len() == 1
        && path.segments.iter().next().unwrap().ident == "Vec"
}

/// Extracts the inner generic type from a `Vec<_>` type.
///
/// For example, given `Vec<String>` this function will return `String`.
#[cfg(feature = "__abi-generate")]
pub(crate) fn extract_vec_type(ty: &Type) -> Option<&Type> {
    match ty {
        Type::Path(type_path) if type_path.qself.is_none() && path_is_vec(&type_path.path) => {
            let type_params = &type_path.path.segments.first()?.arguments;
            let generic_arg = match type_params {
                // We are interested in the first (and only) angle-bracketed param:
                PathArguments::AngleBracketed(params) if params.args.len() == 1 => {
                    Some(params.args.first()?)
                }
                _ => None,
            }?;
            match generic_arg {
                GenericArgument::Type(ty) => Some(ty),
                _ => None,
            }
        }
        _ => None,
    }
}

/// Extracts reference and mutability tokens from a `Type` object. Also, strips top-level lifetime binding if present.
pub(crate) fn extract_ref_mut(ty: &Type) -> syn::Result<(Option<And>, Option<Mut>, Type)> {
    match ty {
        x @ (Type::Array(_) | Type::Path(_) | Type::Tuple(_) | Type::Group(_)) => {
            Ok((None, None, (*x).clone()))
        }
        Type::Reference(r) => Ok((Some(r.and_token), r.mutability, (*r.elem.as_ref()).clone())),
        _ => Err(syn::Error::new_spanned(ty, "Unsupported contract API type.")),
    }
}

/// Checks that the method signature is supported in the NEAR Contract API.
pub(crate) fn sig_is_supported(sig: &Signature) -> syn::Result<()> {
    if sig.asyncness.is_some() {
        return Err(syn::Error::new(sig.span(), "Contract API is not allowed to be async."));
    }
    if sig.abi.is_some() {
        return Err(syn::Error::new(
            sig.span(),
            "Contract API is not allowed to have binary interface.",
        ));
    }
    if sig.variadic.is_some() {
        return Err(syn::Error::new(
            sig.span(),
            "Contract API is not allowed to have variadic arguments.",
        ));
    }

    Ok(())
}

fn _sanitize_self(typ: TokenStream2, replace_with: &TokenStream2) -> (TokenStream2, Vec<Span>) {
    let mut self_occurrences = vec![];
    let trees = typ.into_iter().map(|t| match t {
        TokenTree::Ident(ident) if ident == "Self" => {
            self_occurrences.push(ident.span());
            replace_with
                .clone()
                .into_iter()
                .map(|mut t| {
                    t.set_span(ident.span());
                    t
                })
                .collect::<TokenStream2>()
        }
        TokenTree::Group(group) => {
            let (stream, self_occurrences_inner) = _sanitize_self(group.stream(), replace_with);
            self_occurrences.extend(self_occurrences_inner);
            TokenTree::Group(Group::new(group.delimiter(), stream)).into()
        }
        rest => rest.into(),
    });
    (trees.collect(), self_occurrences)
}

pub fn sanitize_self(typ: &Type, replace_with: &TokenStream2) -> syn::Result<SanitizeSelfResult> {
    let (ty_tokens, self_occurrences) = _sanitize_self(quote! { #typ }, replace_with);

    let ty = syn::parse2(ty_tokens).map_err(|original| {
        syn::Error::new(original.span(), "`Self` sanitization failed. Please report this as a bug.")
    })?;

    Ok(SanitizeSelfResult { ty, self_occurrences })
}

pub struct SanitizeSelfResult {
    pub ty: Type,
    pub self_occurrences: Vec<Span>,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn sanitize_self_works() {
        let typ: Type = syn::parse_str("Self").unwrap();
        let replace_with: TokenStream2 = syn::parse_str("MyType").unwrap();
        let sanitized = sanitize_self(&typ, &replace_with).unwrap().ty;
        assert_eq!(quote! { #sanitized }.to_string(), "MyType");

        let typ: Type = syn::parse_str("Vec<Self>").unwrap();
        let replace_with: TokenStream2 = syn::parse_str("MyType").unwrap();
        let sanitized = sanitize_self(&typ, &replace_with).unwrap().ty;
        assert_eq!(quote! { #sanitized }.to_string(), "Vec < MyType >");

        let typ: Type = syn::parse_str("Vec<Vec<Self>>").unwrap();
        let replace_with: TokenStream2 = syn::parse_str("MyType").unwrap();
        let sanitized = sanitize_self(&typ, &replace_with).unwrap().ty;
        assert_eq!(quote! { #sanitized }.to_string(), "Vec < Vec < MyType > >");

        let typ: Type = syn::parse_str("Option<[(Self, Result<Self, ()>); 2]>").unwrap();
        let replace_with: TokenStream2 = syn::parse_str("MyType").unwrap();
        let sanitized = sanitize_self(&typ, &replace_with).unwrap().ty;
        assert_eq!(
            quote! { #sanitized }.to_string(),
            "Option < [(MyType , Result < MyType , () >) ; 2] >"
        );
    }

    #[test]
    fn sanitize_self_keeps_track_of_replacements() {
        let typ: Type = syn::parse_str("Self").unwrap();
        let replace_with: TokenStream2 = syn::parse_str("MyType").unwrap();
        assert_eq!(sanitize_self(&typ, &replace_with).unwrap().self_occurrences.len(), 1);

        let typ: Type = syn::parse_str("SomeType").unwrap();
        let replace_with: TokenStream2 = syn::parse_str("MyType").unwrap();
        assert!(sanitize_self(&typ, &replace_with).unwrap().self_occurrences.is_empty());

        let typ: Type = syn::parse_str("Option<[(Self, Result<Self, ()>); 2]>").unwrap();
        let replace_with: TokenStream2 = syn::parse_str("MyType").unwrap();
        assert_eq!(sanitize_self(&typ, &replace_with).unwrap().self_occurrences.len(), 2);
    }
}
