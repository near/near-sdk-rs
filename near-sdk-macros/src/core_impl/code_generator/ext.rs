use crate::core_impl::{serializer, AttrSigInfo};
use proc_macro2::{Ident, Span, TokenStream as TokenStream2};
use quote::{format_ident, quote, ToTokens};
use syn::{parse_quote, Attribute, Generics, Path, Signature};

/// Generates inner ext code for structs and modules. If intended for a struct, generic details
/// for the struct should be passed in through `generic_details` and the `ext` method will be
/// added as an impl to the struct ident.
pub(crate) fn generate_ext_structs(
    ident: &Ident,
    generic_details: Option<&Generics>,
) -> proc_macro2::TokenStream {
    let name = format_ident!("{}Ext", ident);
    let mut ext_code = quote! {
        /// API for calling this contract's functions in a subsequent execution.
        pub fn ext(account_id: ::near_sdk::AccountId) -> #name {
            #name {
                account_id,
                deposit: ::near_sdk::NearToken::from_near(0),
                static_gas: ::near_sdk::Gas::from_gas(0),
                gas_weight: ::near_sdk::GasWeight::default(),
            }
        }
    };
    if let Some(generics) = generic_details {
        // If ext generation is on struct, make ext function associated with struct not module
        ext_code = quote! {
            impl #generics #ident #generics {
                #ext_code
            }
        };
    }

    quote! {
      #[must_use]
      pub struct #name {
          pub(crate) account_id: ::near_sdk::AccountId,
          pub(crate) deposit: ::near_sdk::NearToken,
          pub(crate) static_gas: ::near_sdk::Gas,
          pub(crate) gas_weight: ::near_sdk::GasWeight,
      }

      impl #name {
          pub fn with_attached_deposit(mut self, amount: ::near_sdk::NearToken) -> Self {
              self.deposit = amount;
              self
          }
          pub fn with_static_gas(mut self, static_gas: ::near_sdk::Gas) -> Self {
              self.static_gas = static_gas;
              self
          }
          pub fn with_unused_gas_weight(mut self, gas_weight: u64) -> Self {
              self.gas_weight = ::near_sdk::GasWeight(gas_weight);
              self
          }
      }

      #ext_code
    }
}

/// Non-bindgen attributes on contract methods should not be forwarded to the
/// corresponding `_Ext` methods by default. It may lead to compilation errors
/// or unexpected behavior. For a more detailed motivation, see [#959].
///
/// However, some attributes should be forwarded and they are defined here.
///
/// [#959]: https://github.com/near/near-sdk-rs/pull/959
const FN_ATTRIBUTES_TO_FORWARD: [&str; 1] = [
    // Allow some contract methods to be feature gated, for example:
    //
    // ```
    // impl Contract {
    //     #[cfg(integration_tests)]
    //     pub fn test_method(&mut self) { /* ... */ }
    // }
    // ```
    //
    // In that scenario `ContractExt::test_method` should be included only if
    // `integration_tests` is enabled.
    "cfg",
];

/// Returns whether `attribute` should be forwarded to `_Ext` methods, see
/// [`FN_ATTRIBUTES_TO_FORWARD`].
fn is_fn_attribute_to_forward(attribute: &Attribute) -> bool {
    for to_forward in FN_ATTRIBUTES_TO_FORWARD.iter() {
        let to_forward_ident = Ident::new(to_forward, Span::mixed_site());
        let to_forward_path: Path = parse_quote! { #to_forward_ident };
        if &to_forward_path == attribute.meta.path() {
            return true;
        }
    }
    false
}

/// Generate methods on <StructName>Ext to enable calling each method.
pub(crate) fn generate_ext_function_wrappers<'a>(
    ident: &Ident,
    methods: impl IntoIterator<Item = &'a AttrSigInfo>,
) -> TokenStream2 {
    let ext_ident = format_ident!("{}Ext", ident);
    let mut res = TokenStream2::new();
    for method in methods {
        res.extend(generate_ext_function(method));
    }
    quote! {
        impl #ext_ident {
            #res
        }
    }
}

fn generate_ext_function(attr_signature_info: &AttrSigInfo) -> TokenStream2 {
    let pat_type_list = attr_signature_info.pat_type_list();
    let serialize =
        serializer::generate_serializer(attr_signature_info, &attr_signature_info.input_serializer);

    let AttrSigInfo { non_bindgen_attrs, ident, original_sig, .. } = attr_signature_info;
    let ident_str = ident.to_string();
    let mut new_non_bindgen_attrs = TokenStream2::new();
    for attribute in non_bindgen_attrs.iter() {
        if is_fn_attribute_to_forward(attribute) {
            attribute.to_tokens(&mut new_non_bindgen_attrs);
        }
    }
    let Signature { generics, .. } = original_sig;
    quote! {
        #new_non_bindgen_attrs
        pub fn #ident #generics(self, #pat_type_list) -> ::near_sdk::Promise {
            let __args = #serialize;
            ::near_sdk::Promise::new(self.account_id)
            .function_call_weight(
                ::std::string::String::from(#ident_str),
                __args,
                self.deposit,
                self.static_gas,
                self.gas_weight,
            )
        }
    }
}

#[rustfmt::skip]
#[cfg(test)]
mod tests {
    use crate::core_impl::ImplItemMethodInfo;

    use super::*;
    use syn::{parse_quote, ImplItemFn, ItemStruct, Type};
    use crate::core_impl::utils::test_helpers::{local_insta_assert_snapshot, pretty_print_syn_str};

    #[test]
    fn ext_gen() {
        let st: ItemStruct = parse_quote! { struct Test { a: u8 } };
        let actual = generate_ext_structs(&st.ident, Some(&st.generics));
       
        local_insta_assert_snapshot!(pretty_print_syn_str(&actual).unwrap());
    }

    #[test]
    fn module_ext_gen() {
        let ident: Ident = parse_quote! { Test };
        let actual = generate_ext_structs(&ident, None);
    
        local_insta_assert_snapshot!(pretty_print_syn_str(&actual).unwrap());
    }

    /// Verifies that only whitelisted attributes are forwarded to `_Ext`
    /// methods.
    #[test]
    fn ext_fn_non_bindgen_attrs() {
        let impl_type: Type = parse_quote! { Hello };
        let mut method: ImplItemFn = parse_quote! {
            #[cfg(target_os = "linux")]
            #[inline]
            #[warn(unused)]
            pub fn method(&self) { }
        };
        let method_info = ImplItemMethodInfo::new(&mut method, false, impl_type).unwrap().unwrap();
        let actual = generate_ext_function(&method_info.attr_signature_info);

        local_insta_assert_snapshot!(pretty_print_syn_str(&actual).unwrap());
    }

    #[test]
    fn ext_basic_json() {
        let impl_type: Type = parse_quote! { Hello };
        let mut method: ImplItemFn = parse_quote! {
            pub fn method(&self, k: &String) { }
        };
        let method_info = ImplItemMethodInfo::new(&mut method, false, impl_type).unwrap().unwrap();
        let actual = generate_ext_function(&method_info.attr_signature_info);
     
        local_insta_assert_snapshot!(pretty_print_syn_str(&actual).unwrap());
    }

    #[test]
    fn ext_basic_borsh() {
        let impl_type: Type = syn::parse_str("Hello").unwrap();
        let mut method: syn::ImplItemFn = parse_quote! {
          pub fn borsh_test(&mut self, #[serializer(borsh)] a: String) {}
        };
        let method_info = ImplItemMethodInfo::new(&mut method, false, impl_type).unwrap().unwrap();
        let actual = generate_ext_function(&method_info.attr_signature_info);
       
        local_insta_assert_snapshot!(pretty_print_syn_str(&actual).unwrap());
    }
}
