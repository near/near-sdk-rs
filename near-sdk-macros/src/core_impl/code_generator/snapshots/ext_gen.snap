---
source: near-sdk-macros/src/core_impl/code_generator/ext.rs
expression: pretty_print_syn_str(&actual).unwrap()
---
#[must_use]
pub struct TestExt {
    pub(crate) account_id: ::near_sdk::AccountId,
    pub(crate) deposit: ::near_sdk::NearToken,
    pub(crate) static_gas: ::near_sdk::Gas,
    pub(crate) gas_weight: ::near_sdk::GasWeight,
}
impl TestExt {
    pub fn with_attached_deposit(mut self, amount: ::near_sdk::NearToken) -> Self {
        self.deposit = amount;
        self
    }
    pub fn with_static_gas(mut self, static_gas: ::near_sdk::Gas) -> Self {
        self.static_gas = static_gas;
        self
    }
    pub fn with_unused_gas_weight(mut self, gas_weight: u64) -> Self {
        self.gas_weight = ::near_sdk::GasWeight(gas_weight);
        self
    }
}
impl Test {
    /// API for calling this contract's functions in a subsequent execution.
    pub fn ext(account_id: ::near_sdk::AccountId) -> TestExt {
        TestExt {
            account_id,
            deposit: ::near_sdk::NearToken::from_near(0),
            static_gas: ::near_sdk::Gas::from_gas(0),
            gas_weight: ::near_sdk::GasWeight::default(),
        }
    }
}

